accepted/uyuni-microos-salt-ssh.md

```diff
@@ -0,0 +1,85 @@
+
+This poses a problem: on TU systems, Salt needs to coordinate files between the IAT and OOT filesystems.
+
+## Solution
+
+There are two Salt modules that modify the filesystem:
+
+- `state` (e.g. `state.apply`)
+- `transactional_update` (e.g. `transactional_update.apply`)
+
+Users should use the `state` module modify OOT filesystem, and the `transactional_update` module to modify the IAT filesystem.
+
+Because `state` module will not use transactions, it requires no changes for Salt SSH to work.
+However, the `transactional_update` module requires:
+
+1. a Salt SSH wrapper to prepare and transfer the `salt_state.tgz` file
+2. to implement the `pkg` function, which executes the `salt_state.tgz` file
+3. to implement a synchronization mechanism that copies Salt and the `salt_state.tgz` file into the transaction
+
+Steps 1. and 2. for the `transactional_module` are equivalent to the `state` module:
+
+1. [state's SSH wrapper](https://github.com/openSUSE/salt/blob/openSUSE/release/3006.0/salt/client/ssh/wrapper/state.py)
+2. `transactional_module` can delegate execution to `state.pkg` once in a transaction with all the necessary files
+
+For step 3, this RFC proposes the following mechanism:
+
+- When `transactional_update.pkg` function is called, it first deploys itself from whatever location Salt initially is on a minion (e.g. `/var/tmp/.root_12345_salt`) to `/var/cache`.
+  - In case of a traditional Salt, `transactional_update` generates a thin Salt client into `/var/cache`.
+  - In case of a Salt Bundle (`venv-salt-minion`), `transactional_update` copies the directory from which Salt is currently executed, e.g. `/var/tmp/venv-salt-minion` to `/var/cache`. Consequently, `transactional_update` must also fix the shebangs in the new `/var/cache` copy.
+
+> [!NOTE]
+> In TU systems, the `/var/cache` location is shared between the IAT and OOT filesystems, which is why it serves as a synchronization point.
+
+- After deploying a copy of Salt into `/var/cache`, `transactional_update` also copies the `salt_state.tgz` file to `/var/cache`.
+- Then, the Salt process outside of a transaction (_Salt_OOT_) starts a new Salt process inside of a transaction (_Salt_IAT_), which now executes `state.pkg`.
+- After the Salt_IAT process finishes, it passes the execution result to Salt_OOT.
+- Salt_OOT deletes the Salt copy from the `/var/cache` location, and returns execution result to Salt master.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Greater resource requirements: each execution of Salt SSH deploys a Salt client to the temporary `/var/cache` location, executes a new Salt process, and removing the temporary Salt client at the end of the execution. This means increased storage footprint (both RAM and disk), as well as added CPU cycles related to copying and cleaning up Salt.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Not supporting Salt SSH on TU systems.
```

Comment:
Instead of "not supporting" salt ssh on TU systems, could we instead suggest using `transactional_update` execution module? Or can we even provide a transactional_update state which would wrap TU execution module?

i.e. instead of `pkg.installed` suggest using  `transaction_pkg.installed`?

Of course this would open another can of worms like:
* users required to keep in mind mandatory reboot
* detect we are already running in the transaction via tu executor in normal salt

Reply:
I'm not sure I understand here. This RFC is about requiring the users to use the `state` or `transactional_update` modules based on their needs:

> Users should use the `state` module modify OOT filesystem, and the `transactional_update` module to modify the IAT filesystem.

So if I'm not misunderstanding, we already suggest that user use the `transactional_update` when necessary. This leads us to changing how Uyuni/SUMA deals with TU minions as well.

Regarding:

> i.e. instead of pkg.installed suggest using transaction_pkg.installed?

It would work for some changes, e.g. changes that do not require other files, like the `pkg.installed` you mention. It would not work for changes that require anything on the filesystem, e.g.:

- modules that execute `sls` files
-  the `file` command that copies a file to a different location, etc.

This would require the same synchronization mechanism as we provide in this RFC. 

I can add it to alternative solutions with the caveat that we'd support only a subset of Uyuni's functionality (and of course, we'd also have to re-do how we treat TU minions on Uyuni's side since neither `state.apply` nor `transactional_update.apply` would work)